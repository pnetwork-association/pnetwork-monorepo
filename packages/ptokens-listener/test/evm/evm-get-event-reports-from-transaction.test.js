const ethers = require('ethers')
const constants = require('ptokens-constants')
const { receipts } = require('../mock/evm-receipts')

describe('Get EVM event reports', () => {
  describe('getEvmEventReportsFromTransaction', () => {
    beforeEach(() => {
      jest.resetAllMocks()
    })
    afterEach(() => {
      jest.useRealTimers()
    })
    it('Should get event reports for a transaction', async () => {
      jest.useFakeTimers({ now: Date.parse('2023-05-18T07:35:54.575Z') })
      const getTransactionReceiptSpy = jest
        .spyOn(ethers.providers.JsonRpcProvider.prototype, 'getTransactionReceipt')
        .mockResolvedValue(receipts[0])
      const fakeProvider = {
        getTransactionReceipt: getTransactionReceiptSpy,
      }
      const getDefaultProviderSpy = jest
        .spyOn(ethers.providers, 'JsonRpcProvider')
        .mockImplementation(_url => fakeProvider)

      const {
        getEvmEventReportsFromTransaction,
      } = require('../../lib/evm/evm-get-event-reports-from-transaction')
      const provider = 'polygon-provider-url-1'
      const networkId = constants.networkIds.POLYGON_MAINNET
      const txHash = '0xa5c5838123aa37d2efd69285f7b6bd8c2e93d4cf243d45926169502c13b23a49' // secretlint-disable-line
      const ret = await getEvmEventReportsFromTransaction(
        provider,
        networkId,
        txHash,
        constants.evm.events.OPERATION_EXECUTED_SIGNATURE
      )
      expect(getDefaultProviderSpy).toHaveBeenNthCalledWith(1, provider)
      expect(getTransactionReceiptSpy).toHaveBeenNthCalledWith(1, txHash)
      expect(ret).toMatchSnapshot()
    })
  })
})
