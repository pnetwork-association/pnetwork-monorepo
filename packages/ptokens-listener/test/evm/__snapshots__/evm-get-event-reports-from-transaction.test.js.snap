// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get EVM event reports getEvmEventReportsFromTransaction Should get event reports for a transaction 1`] = `
[
  {
    "_id": "operationexecuted_0xf9b459a1_0xa5c5838123aa37d2efd69285f7b6bd8c2e93d4cf243d45926169502c13b23a49",
    "assetAmount": "200000",
    "assetTokenAddress": null,
    "blockHash": "0x0fc3588f727dde10ccd937b04f5666fb04e39553b4c73719555acd7a6a430764",
    "destinationAccount": "0xdDb5f4535123DAa5aE343c24006F4075aBAF5F7B",
    "destinationNetworkId": "0xf9b459a1",
    "eventArgs": [
      [
        "0x05cf0e83408207704ee0ea2a4a6ea87905fc0d2038dbb610a0ca64f2cf47b134",
        "0xb1bb8b6502edc17fdd0cc83505289a6d429a6381ffe5dbf4fe31a88dd236d643",
        "0x0000000000000000000000000000000000000000000000000000000000000000",
        "98322",
        "18",
        "200000",
        "0",
        "1000",
        "2000",
        "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
        "0x5aca268b",
        "0xf9b459a1",
        "0xb9286154",
        "0x5aca268b",
        "0xddb5f4535123daa5ae343c24006f4075abaf5f7b",
        "0xdDb5f4535123DAa5aE343c24006F4075aBAF5F7B",
        "pNetwork Token",
        "PNT",
        "0x",
        false,
      ],
    ],
    "eventName": "OperationExecuted",
    "finalTransactionHash": null,
    "finalTransactionTimestamp": null,
    "forwardDestinationNetworkId": "0xb9286154",
    "forwardNetworkFeeAssetAmount": "2000",
    "isForProtocol": false,
    "networkFeeAssetAmount": "1000",
    "networkId": "0xf9b459a1",
    "nonce": "98322",
    "operationId": "0xd9feb6e60cd73c396cbaeb3e5fa55c774c03a274c54f5bc53a62a59855ec7cc4",
    "optionsMask": "0x0000000000000000000000000000000000000000000000000000000000000000",
    "originatingAddress": "0xddb5f4535123daa5ae343c24006f4075abaf5f7b",
    "originatingBlockHash": "0x05cf0e83408207704ee0ea2a4a6ea87905fc0d2038dbb610a0ca64f2cf47b134",
    "originatingNetworkId": "0x5aca268b",
    "originatingTransactionHash": "0xb1bb8b6502edc17fdd0cc83505289a6d429a6381ffe5dbf4fe31a88dd236d643",
    "proposedTransactionHash": null,
    "proposedTransactionTimestamp": null,
    "status": "detected",
    "transactionHash": "0xa5c5838123aa37d2efd69285f7b6bd8c2e93d4cf243d45926169502c13b23a49",
    "underlyingAssetDecimals": 18,
    "underlyingAssetName": "pNetwork Token",
    "underlyingAssetNetworkId": "0x5aca268b",
    "underlyingAssetSymbol": "PNT",
    "underlyingAssetTokenAddress": "0xdaacB0Ab6Fb34d24E8a67BfA14BF4D95D4C7aF92",
    "userData": "0x",
    "userDataProtocolFeeAssetAmount": "0",
    "witnessedTimestamp": "2023-05-18T07:35:54.575Z",
  },
]
`;
